src/collective/communicator.cu

```diff
@@ -0,0 +1,40 @@
+/*!
+ * Copyright 2022 XGBoost contributors
+ */
+#include "communicator.h"
+#include "device_communicator.cuh"
+#include "device_communicator_adapter.cuh"
+#ifdef XGBOOST_USE_NCCL
+#include "nccl_device_communicator.cuh"
+#endif
+
+namespace xgboost {
+namespace collective {
+
+thread_local int Communicator::device_ordinal_{-1};
```

Comment:
We want to do something for this parameter, could you please take a look at https://github.com/dmlc/xgboost/issues/7308 ? One issue in my experimental implementation is that we need to consider multi-threaded environments like Python async functions. In such environments, it's difficult to enforce synchronization of these thread local parameters.

Reply:
The proposed python interface in #7308 looks good.

Here the device id doesn't have to be per thread, assuming we have a global configuration that sets it, and it's available when we call `Communicator::Init()`. But the communicators do need to be thread local since they are not thread safe.